File: commands/create.py
import argparse
import os
import unicodedata


def display_width(text):
    width = 0
    for ch in text:
        eaw = unicodedata.east_asian_width(ch)
        width += 2 if eaw in ("W", "F") else 1
    return width


def create_note(name, directory="."):
    file_name = f"{name}.txt"
    file_path = os.path.join(directory, file_name)

    title = name
    title_underline = "=" * display_width(title)

    # Support `note`
    if name.endswith("Note"):
        section = name[:-5].rstrip()
        section_underline = "-" * display_width(section)
        section_block = f"{section}\n{section_underline}\n\n\n"
    else:
        section_block = ""

    content = f"\n" f"{title}\n" f"{title_underline}\n" f"\n" f"\n" f"{section_block}"

    with open(file_path, "w", encoding="UTF8") as f:
        f.write(content)

    print(f"Created: {file_path}")


def build_parser():
    parser = argparse.ArgumentParser(description="Create a new note file.")
    parser.add_argument(
        "name",
        help="Basename of the note file (e.g. 'ABC Note' creates 'ABC Note.txt').",
    )
    parser.add_argument(
        "--directory",
        "-d",
        default=".",
        help="Directory to create the note in. Defaults to current directory.",
    )
    return parser


def main(argv=None):
    parser = build_parser()
    args = parser.parse_args(argv)

    create_note(args.name, args.directory)


if __name__ == "__main__":
    main()

File: rust/src/commands/create.rs
use std::fs;
use std::path::PathBuf;
use unicode_width::UnicodeWidthStr;

fn display_width(text: &str) -> usize {
    text.width()
}

pub fn run(name: &str, directory: &str) -> Result<(), String> {
    let file_name = format!("{}.txt", name);
    let file_path = PathBuf::from(directory).join(&file_name);

    let title = name;
    let title_underline = "=".repeat(display_width(title));

    // Support `note`
    let section_block = if name.ends_with("Note") {
        let section = name[..name.len() - 4].trim_end();
        let section_underline = "-".repeat(display_width(section));
        format!("{}\n{}\n\n\n", section, section_underline)
    } else {
        String::new()
    };

    let content = format!(
        "\n{}\n{}\n\n\n{}",
        title, title_underline, section_block
    );

    fs::write(&file_path, content)
        .map_err(|e| format!("failed to write '{}': {}", file_path.display(), e))?;

    println!("Created: {}", file_path.display());
    Ok(())
}

pub fn main(argv: &[String]) {
    let mut name: Option<String> = None;
    let mut directory = ".".to_string();

    let mut i = 0usize;
    while i < argv.len() {
        match argv[i].as_str() {
            "--directory" | "-d" => {
                if i + 1 < argv.len() {
                    directory = argv[i + 1].clone();
                    i += 2;
                } else {
                    eprintln!("Error: --directory requires an argument.");
                    std::process::exit(1);
                }
            }
            arg if !arg.starts_with('-') => {
                name = Some(arg.to_string());
                i += 1;
            }
            arg => {
                eprintln!("Error: unrecognized argument '{}'.", arg);
                std::process::exit(1);
            }
        }
    }

    let name = match name {
        Some(n) => n,
        None => {
            eprintln!("Error: no name provided.");
            std::process::exit(1);
        }
    };

    if let Err(e) = run(&name, &directory) {
        eprintln!("Error: {}", e);
        std::process::exit(1);
    }
}
